keras/src/metrics/accuracy_metrics.py

```diff
@@ -9,10 +9,7 @@ def accuracy(y_true, y_pred):
     y_pred = ops.convert_to_tensor(y_pred)
     y_true = ops.convert_to_tensor(y_true, dtype=y_pred.dtype)
     y_true, y_pred = squeeze_or_expand_to_same_rank(y_true, y_pred)
-    return ops.mean(
-        ops.cast(ops.equal(y_true, y_pred), dtype=backend.floatx()),
-        axis=-1,
-    )
+    return ops.cast(ops.equal(y_true, y_pred), dtype=backend.floatx())
```

Comment:
`accuracy` and `binary_accuracy` should reduce on the last dimension at this stage (unless it the input has rank 1, in which case the last dimension is the batch dimension). All metrics functions do this.

So if your inputs have rank 2, then the function will return one accuracy scalar score per entry in the batch (e.g. `(32, 4) -> (32,)`)

Reply:
Yes, the reduction is applied in a generic way here (if weights have a smaller rank):
https://github.com/keras-team/keras/blob/master/keras/src/metrics/reduction_metrics.py#L29-L31
and/or here:
https://github.com/keras-team/keras/blob/master/keras/src/metrics/reduction_metrics.py#L39-L41
So it does work as expected for rank 2 and above.

Let me make sure it works as expected for rank 1 `y`s and add unit tests.

The issue is that we don't prevent weights that are the same rank as `y_true` and `y_pred`. What it was doing before was multiplying the average of the equality by the average of the weights. Now it's averaging the equality multiplied by the weights.

The full sequence is now:
1. compute equality (called `values`)
2. tweak `values` and `sample_weights` to get to the same rank, which can involve a `mean` on the last few dimensions of `values`
3. multiply `values` by `sample_weights`
4. mean reduction of `values` and `sample_weights` to rank 1
5. sum `values` (`total`) and `sample_weights` (`num_samples`)
6. divide `total` by `num_samples`.

Before, there was an extra `mean` between 1. and 2. It did not matter if `rank(sample_weights) < rank(values)` because step 2. does a `mean` to lower the rank of `values`.

Note that it now follows the same pattern as `tf_keras` where the `accuracy` function is not the one doing the reduction: https://github.com/keras-team/tf-keras/blob/master/tf_keras/metrics/accuracy_metrics.py#L354-L364
but the reduction is done generically in the `Reduce` class: https://github.com/keras-team/tf-keras/blob/master/tf_keras/metrics/base_metric.py#L520-L522 and https://github.com/keras-team/tf-keras/blob/master/tf_keras/metrics/base_metric.py#L525